File: src/leetCodeProblems/StacksAndQueues/MinStack155.java
package leetCodeProblems.StacksAndQueues;

/**
 * LeetCode - https://leetcode.com/problems/min-stack/
 *
 * TimeComplexity - O(1)
 */

import java.util.Stack;

public class MinStack155 {

    Stack<Integer> mainStack;
    Stack<Integer> monotonicStack;

    public MinStack155() {
        mainStack = new Stack<>();
        monotonicStack = new Stack<>();
    }

    public void push(int val) {
        if (!monotonicStack.isEmpty()) {
            if (monotonicStack.peek() >= val) {
                monotonicStack.push(val);
            }
        }
        else {
            monotonicStack.push(val);
        }

        //System.out.println(monotonicStack);
        mainStack.push(val);
    }

    public void pop() {
        int element = mainStack.pop();

        if (!monotonicStack.isEmpty()) {
            if (monotonicStack.peek() == element) {
                monotonicStack.pop();
            }
        }

        return;
    }

    public int top() {
        return mainStack.peek();
    }

    public int getMin() {
        if (monotonicStack.isEmpty()) {
            return -1;
        }

        return monotonicStack.peek();
    }
}

File: src/leetCodeProblems/MonotonicStack/MinStack155.go
package main

/*
- Leetcode - https://leetcode.com/problems/min-stack/description/
- Time - O(1)
- Space - O(n)
*/

type stack3 []int

func (s *stack3) push(n int) {
	*s = append(*s, n)
}

func (s *stack3) pop() int {
	temp := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]
	return temp
}

func (s stack3) isEmpty() bool {
	return len(s) == 0
}

func (s stack3) peek() int {
	return s[len(s)-1]
}

type MinStack struct {
	s  *stack3 // main stack to store all values
	ms *stack3 // auxiliary stack to store minimum values
}

func Constructor1() MinStack {
	return MinStack{
		s:  new(stack3),
		ms: new(stack3),
	}
}

// Push adds a new value to the MinStack
// If the value is smaller than or equal to the current minimum,
// it is also pushed onto the ms (min stack)
func (this *MinStack) Push(val int) {
	if this.ms.isEmpty() || this.ms.peek() >= val {
		this.ms.push(val)
	}

	this.s.push(val)
}

// Pop removes the top element from the main stack,
// and also removes it from the min stack if it was the minimum
func (this *MinStack) Pop() {
	if !this.ms.isEmpty() &&
		this.s.peek() == this.ms.peek() {
		this.ms.pop()
	}

	this.s.pop()
}

func (this *MinStack) Top() int {
	return this.s.peek()
}

func (this *MinStack) GetMin() int {
	return this.ms.peek()
}

// func main() {
// 	min := Constructor1()
// 	min.Push(-2)
// 	min.Push(0)
// 	min.Push(-3)
// 	log.Println(min.GetMin())
// 	min.Pop()
// 	log.Println(min.Top())
// 	log.Println(min.GetMin())
// }
